api/app/serializers/spree/v2/storefront/product_serializer.rb

```diff
@@ -47,7 +47,10 @@ class ProductSerializer < BaseSerializer
         has_many :variants
         has_many :option_types
         has_many :product_properties
-        has_many :taxons
+
+        has_many :taxons do |object, params|
+          params[:store].taxons.joins(:products).where(products: { id: object.id })
```

Comment:
```suggestion
          object.taxons.for_store(params[:store])
```

let's move this into the model as we're going to re-use this :) 

in Taxon model:

```ruby
scope :for_store, ->(store) { joins(:taxonomy).where(spree_taxonomies: { store_id: store.id }) }
```

Reply:
@damianlegawiec I can do that, but it is overriding default `for_store` scope. https://github.com/spree/spree/blob/master/core/app/models/spree/base.rb#L25

Reply:
@szymoniwacz sorry my bad, we don't need it :)